Fix enum attributes are not members

terencehonles opened the pull request:
This adds on to the change in https://github.com/python/mypy/pull/17182 and fixes enum attributes being used as members.

@hamdanal / @hauntsaninja  I noticed there was the function `get_enum_values` in mypy types and instead of adding more to the case in `try_expanding_sum_type_to_union` it seemed like it might make sense to move the changes from  https://github.com/python/mypy/pull/17182 there. I don't think the other code touched in that PR can use `get_enum_values`.

fixes: https://github.com/python/mypy/issues/16730

<!--
Checklist:
- Read the [Contributing Guidelines](https://github.com/python/mypy/blob/master/CONTRIBUTING.md)
- Add tests for all changed behaviour.
- If you can't add a test, please explain why and how you verified your changes work.
- Make sure CI passes.
- Please do not force push to the PR once it has been reviewed.
-->

hamdanal commented on mypy/types.py:
```diff
@@ -1496,7 +1496,14 @@ def is_singleton_type(self) -> bool:
+                isinstance(sym.node, mypy.nodes.Var)
+                and name not in ENUM_REMOVED_PROPS
+                and not name.startswith("__")
+                and sym.node.has_explicit_value
```
This last condition changes the behavior of mypy which currently considers unassigned but annotated attributes as members of the enum.
I would say the new behavior is better but it must be noted in the PR title or description that this is a breaking change that must be considered on its own.
Looking at the mypy primer output, it looks like members of enums constructed using the enum call syntax do not have the `has_explicit_value` set which breaks type narrowing. I suggest changing the variable [here](https://github.com/python/mypy/blob/fb31409b392c5533b25173705d62ed385ee39cfb/mypy/semanal_enum.py#L143-L146) to also set this flag perhaps with a comment explaining the reason -- something like "`E = Enum("E", "A")` is equivalent to `class E(Enum): A = auto()`"

terencehonles commented on test-data/unit/check-python310.test:
```diff
@@ -1484,6 +1484,60 @@ def g(m: Medal) -> int:
+    def __new__(cls, value: int): pass
+
+class Medal(int, Enum):
+    prize: str
```
yup, should be able to tomorrow

terencehonles replied:
I thought about this a bit, and this is either a request for a `nonmember(...)` test, which is currently not supported, or an enum like the one above. I've added a test for `nonmember` that's been marked as expected to fail: https://github.com/python/mypy/pull/17207/files#diff-4f8d4b202949b4cf2221e3295fbbd1774e152e44b141880981895c349599e927R1-R36

Let me know if that's what you wanted, and/or if we should keep that

terencehonles replied:
A class attribute that's not declared as `nonmember` is an enum member is it not? The case below is a different type, but seems like what you were suggesting if you didn't mean `nonmember`

```python
class Medal(Enum):
    gold = 1
    silver = 2
    bronze = 3
```

Or did you want one that declared a type? I don't remember if I saw one with a type, but if that's all you meant I can add that. Otherwise I'm going to need some guidance to what you're requesting.

terencehonles replied:
Or did you want one that also implements `__new__`?

JelleZijlstra replied:
I specifically want an *annotated* attribute. The specification says that should be a type checker error: https://typing.readthedocs.io/en/latest/spec/enums.html#defining-members

terencehonles replied:
I'm sorry, I was even thinking that seems odd to put typing on an attribute, but I guess I didn't realize you wanted a failing test case (and I didn't realize it was in the spec, I just didn't think people would want to do that for probably the very reason it's declared :sweat_smile: )

terencehonles replied:
I was looking at adding the test and was also looking to see if anything matched that test case already (just in case it did). I came across https://github.com/python/mypy/blob/740d39ecc31c64675439e6796dcf46e6a9110896/test-data/unit/check-enum.test#L1755-L1781 which is actually a test case that was written to support a change that fixed an issue _I_ reported. Go figure!

It seems like this is currently not a type checker error and I can look at possibly adding it, but I will have to look either over the weekend or next week. I can't really write an xfail test since I'm not really sure what the error message would be (I guess it wouldn't match a not failed test so the error message might not be that important).

Just calling out that I did look into this, but will not be able to complete it immediately

hamdanal replied:
I think what Jelle is asking for is something like this (you'll also need to revert the changes to check-python311.test):
```patch
diff --git a/mypy/semanal.py b/mypy/semanal.py
index d2f02d483..7f23904e6 100644
--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -4200,6 +4200,8 @@ class SemanticAnalyzer(
                     ),
                     lvalue,
                 )
+            if explicit_type and has_explicit_value:
+                self.fail("Type annotations are not allowed for enum members", lvalue)

         if (not existing or isinstance(existing.node, PlaceholderNode)) and not outer:
             # Define new variable.
diff --git a/test-data/unit/check-enum.test b/test-data/unit/check-enum.test
index d53935085..0f41865ad 100644
--- a/test-data/unit/check-enum.test
+++ b/test-data/unit/check-enum.test
@@ -1764,7 +1764,7 @@ class B(A):
     x = 1  # E: Cannot override writable attribute "x" with a final one

 class A1(Enum):
-    x: int = 1
+    x: int = 1  # E: Type annotations are not allowed for enum members
 class B1(A1):  # E: Cannot extend enum with existing members: "A1"
     pass

@@ -2185,3 +2185,12 @@ reveal_type(A.y.value)  # N: Revealed type is "Literal[2]?"
 def some_a(a: A):
     reveal_type(a.value)  # N: Revealed type is "Union[Literal[1]?, Literal[2]?]"
 [builtins fixtures/dict.pyi]
+
+[case testErrorOnAnnotatedMember]
+from enum import Enum
+
+class Medal(Enum):
+    gold: int = 1  # E: Type annotations are not allowed for enum members
+    silver: str = 2  # E: Type annotations are not allowed for enum members \
+                     # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+    bronze = 3
```

JelleZijlstra replied:
Thanks! I primarily want a test case to make sure this case is covered by the tests. Ideally of course we should follow the spec and raise an error, but that can wait for a later PR too.

terencehonles replied:
Thanks @hamdanal, I finally understood the request, but I'll check out the implementation you provided! I was just headed to sleep and didn't know if I'd have time this weekend to address the issue so I was just following up! I'll try to get to this in the next couple days

terencehonles replied:
updated

terencehonles commented:
@hamdanal sorry for the delay getting back to this, I went on vacation and then coming back was pretty busy. Work's on break for the week so I figured I could pick this up to keep it from getting delayed any further.

I was having trouble forcing mypy to not use the explicitly provided annotation for an enum member and I will probably have to leave it at this.

I left a comment https://github.com/python/mypy/pull/17207/files#diff-8b48b1eca587106e3806bfa9e14b7f7f344e117b203ffc716ce6fef7f2e68fe6R1610 about the test that was added for  https://github.com/python/mypy/issues/11971. [The code in question](https://gitlab.com/cobib/cobib/-/blob/efda494e568268e9e002e76cfaeb1d740f6dce64/src/cobib/config/event.py#L119) will need to ignore this new error, but as mypy better supports enums the code will need to change as it seems like it's abusing enums a bit.

terencehonles commented:
@hamdanal / @JelleZijlstra friendly ping if you have a chance to look over the PR or suggestions on how to fix ^^^

terencehonles commented:
@hamdanal / @JelleZijlstra friendly ping

hamdanal reviewed (approved):
LGTM, thanks.

hauntsaninja reviewed (approved):
Thanks! This change brings mypy closer in line with the new typing spec chapter on enums

hauntsaninja commented:
I'm confused by the primer hits on Tanjun

JelleZijlstra commented:
> I'm confused by the primer hits on Tanjun

I looked a little and couldn't figure out what's going on there either. Both cases involve narrowing an enum attribute inside a loop. I wasn't able to reproduce a similar error with a simple test case, or even by cloning Tanjun and running mypy from this PR directly on those files.

hauntsaninja commented:
I wasn't able to repro minimised, but I can repro locally via `mypy_primer --repo https://github.com/hauntsaninja/mypy --new 46527644f --old 7f3d7f8f1 -k Tanjun`. Note that `--warn-unreachable` is off by default

hauntsaninja commented:
Oh wait, it vendors inspect

hauntsaninja commented:
This is the regression:
```
λ cat test10.py
import enum
from typing import ClassVar, Literal, cast

class _ParameterKind(enum.IntEnum):
    POSITIONAL_ONLY: int
    POSITIONAL_OR_KEYWORD: int
    VAR_POSITIONAL: int
    KEYWORD_ONLY: int
    VAR_KEYWORD: int

class Parameter:

    POSITIONAL_ONLY: ClassVar[Literal[_ParameterKind.POSITIONAL_ONLY]]
    POSITIONAL_OR_KEYWORD: ClassVar[Literal[_ParameterKind.POSITIONAL_OR_KEYWORD]]
    VAR_POSITIONAL: ClassVar[Literal[_ParameterKind.VAR_POSITIONAL]]
    KEYWORD_ONLY: ClassVar[Literal[_ParameterKind.KEYWORD_ONLY]]
    VAR_KEYWORD: ClassVar[Literal[_ParameterKind.VAR_KEYWORD]]
    @property
    def kind(self) -> _ParameterKind: ...


def get_kwargs() -> None:
    parameter = cast(Parameter, ...)

    reveal_type(parameter)
    reveal_type(parameter.kind)

    if parameter.kind is parameter.VAR_KEYWORD:
        return

    print("should be reachable")

λ mypy test10.py --warn-unreachable --strict --disable-error-code empty-body
test10.py:25: note: Revealed type is "test10.Parameter"
test10.py:26: note: Revealed type is "test10._ParameterKind"
test10.py:31: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)
``

terencehonles commented:
@hauntsaninja I believe this might be related to my comment:

> I was having trouble forcing mypy to not use the explicitly provided annotation for an enum member and I will probably have to leave it at this.

But this seems to also not be warning about the:

```python
class _ParameterKind(enum.IntEnum):
    POSITIONAL_ONLY: int
    POSITIONAL_OR_KEYWORD: int
    VAR_POSITIONAL: int
    KEYWORD_ONLY: int
    VAR_KEYWORD: int
```

and I can minimally reproduce with:

```
[case testEnumWithAnnotationOnly]                                              
# flags: --warn-unreachable                                                    
import enum                                                                    
                                                                               
                                                                               
class E(enum.IntEnum):                                                         
    A: int                                                                     
    B: int                                                                     
                                                                               
                                                                               
def do_check(value: E) -> None:                                                
    reveal_type(value)  # N: Revealed type is "__main__.E"                     
    if value is E.A:                                                           
        return                                                                 
                                                                               
    reveal_type(value)  # N: Revealed type is "Literal[__main__.E.B]"          
    "should be reachable"                                                      
                                                                               
[builtins fixtures/primitives.pyi]                                             
[typing fixtures/typing-full.pyi] 
```

So maybe it's actually that? If I add a `A = auto(); B = auto()` it does then warn and is not unreachable anymore. I'm adding a test (unfixed) to make sure an enum without values also warns about being annotated.

terencehonles commented:
@hauntsaninja I got to the bottom of it, and I fixed it in a way that I figured would have the least amount of knock on effects. According to the spec `_ParameterKind` should not be warning, but it should be treated as having no members. Which is what I fixed here https://github.com/python/mypy/pull/17207#discussion_r1590072571

The issue though is that `try_expanding_sum_type_to_union` code would return `Never` for an enum with no members, and because `Never` is also effectively the base class of all types `Never` *is* `_ParameterKind.VAR_KEYWORD`, and therefore mypy is "correctly" warning that the code is not reachable. I don't think `Never` should be treated that way, but I don't want to break other things, so instead I have the `try_expanding_sum_type_to_union` code check how many members it expands to, and if there are no members it doesn't expand the type and leaves it as is. This may need to stay regardless, but I think the `Never is Any` may cause other subtle bugs.

terencehonles commented:
> This may need to stay regardless, but I think the `Never is Any` may cause other subtle bugs.

Actually I think if there was a warning if an expression is always true or always false, then this should be removed since the way I implemented things is that it hides `Never` and that detection would be suppressed here.

hauntsaninja reviewed (approved):
Thanks, the fix looks great!

I think it might be worth issuing a warning for enums with no members, but only in type stubs. I'll make a PR for that.
